migrations.go

```diff
@@ -17,9 +17,28 @@ func runDatabaseMigrations(db *gorm.DB) error {
 		return fmt.Errorf("migrating without constraints: %w", err)
 	}
 
-	db.DisableForeignKeyConstraintWhenMigrating = false
-	if err := db.AutoMigrate(tables...); err != nil {
-		return fmt.Errorf("migrating with constraints: %w", err)
+	if err := db.Transaction(func(tx *gorm.DB) error {
+		// since v4.2.1, drop these constraints to refresh the constraint behavior.
+		// Previously it was RESTRICT, now it's CASCADE.
+		if err := dropOldConstraints(tx, []constraintToDrop{
+			{"artifact", "fk_artifact_build"},
+			{"log", "fk_log_build"},
+			{"build", "fk_build_project"},
+			{"log", "fk_log_build"},
+			{"branch", "fk_project_branches"},
+			{"build_param", "fk_build_params"},
+		}); err != nil {
+			return err
+		}
+
+		tx.DisableForeignKeyConstraintWhenMigrating = false
+		if err := tx.AutoMigrate(tables...); err != nil {
+			return fmt.Errorf("migrating with constraints: %w", err)
+		}
+
+		return nil
+	}); err != nil {
+		return fmt.Errorf("transaction failed: %w", err)
```

Comment:
nit: To not produce double-phrasing in the error message, I think it's safe to just return the error as-is here.

In my mind, at least, it's the job of `db.Transaction` to say that `db.Transaction` failed. If we are going to extend the error message then we should supply it with context that itself would not be able to provide.

As the code paths inside the transaction func all return pretty detailed errors then I vote for just doing this:

```suggestion
		return err
```